packages/apollo-link-error/src/index.ts

```diff
@@ -30,36 +31,67 @@ export const onError = (errorHandler: ErrorHandler): ApolloLink => {
   return new ApolloLink((operation, forward) => {
     return new Observable(observer => {
       let sub;
+      let retrySub;
+      let retrying = false;
+
       try {
         sub = forward(operation).subscribe({
           next: result => {
             if (result.errors) {
-              errorHandler({
+              const retryForward = errorHandler({
                 graphQLErrors: result.errors,
                 response: result,
                 operation,
+                forward,
               });
+
+              if (retryForward) {
+                retrying = true;
+                retrySub = retryForward.subscribe({
+                  next: result => observer.next(result),
+                  error: networkError => observer.error(networkError),
+                  complete: observer.complete.bind(observer),
+                });
+                return;
+              }
             }
             observer.next(result);
           },
           error: networkError => {
-            errorHandler({
+            const retryForward = errorHandler({
               operation,
               networkError,
               //Network errors can return GraphQL errors on for example a 403
               graphQLErrors: networkError.result && networkError.result.errors,
+              forward,
             });
+            if (retryForward) {
+              retrying = true;
+              retrySub = retryForward.subscribe({
+                next: result => observer.next(result),
+                error: networkError => observer.error(networkError),
+                complete: observer.complete.bind(observer),
+              });
+              return;
+            }
             observer.error(networkError);
           },
-          complete: observer.complete.bind(observer),
+          complete: () => {
+            // disable the previous sub from calling complete on observable
+            // if retry is in flight.
+            if (!retrying) {
```

Comment:
nice catch! Can we reuse `retrySubscription` instead of defining another variable?

Reply:
fixed